src/coredump/coredump.c

```diff
@@ -379,92 +464,120 @@ static int save_external_coredump(
         if (fd < 0)
                 return log_error_errno(fd, "Failed to create temporary file for coredump %s: %m", fn);
 
-        r = copy_bytes(input_fd, fd, max_size, 0);
-        if (r < 0) {
-                log_error_errno(r, "Cannot store coredump of %s (%s): %m",
-                                context->meta[META_ARGV_PID], context->meta[META_COMM]);
-                goto fail;
-        }
-        *ret_truncated = r == 1;
-        if (*ret_truncated)
-                log_struct(LOG_INFO,
-                           LOG_MESSAGE("Core file was truncated to %zu bytes.", max_size),
-                           "SIZE_LIMIT=%zu", max_size,
-                           "MESSAGE_ID=" SD_MESSAGE_TRUNCATED_CORE_STR);
+        /* If storage is on tmpfs, the kernel oomd might kill us if there's MemoryMax set on
+         * the service or the slice it belongs to. This is common on low-resources systems,
+         * to avoid crashing processes to take away too many system resources.
+         * Check the cgroup settings, and set max_size to a bit less than half of the
+         * available memory left to the process.
+         * Then, attempt to write the core file uncompressed first - if the write gets
+         * interrupted, we know we won't be able to write it all, so instead compress what
+         * was written so far, delete the uncompressed truncated core, and then continue
+         * compressing from STDIN. Given the compressed core cannot be larger than the
+         * uncompressed one, and 1KB for metadata is accounted for in the calculation, we
+         * should be able to at least store che full compressed core file. */
+
+        storage_on_tmpfs = fd_is_temporary_fs(fd);
+        if (storage_on_tmpfs && arg_compress) {
+                uint64_t cgroup_limit = UINT64_MAX;
+
+                /* If we can't get the cgroup limit, just ignore it, but don't fail,
+                 * try anyway with the config settings. */
+                r = get_cgroup_memory_limit(&cgroup_limit);
+                if (r < 0)
+                        log_debug_errno(r, "Failed to get current cgroup memory limit, "
+                                           "fallback to static config settings: %m");
```

Comment:
should probably also take free space on the fs as reported by statfs()/sfatvfs() into account?

Reply:
well, sure, but if people have only 5M left on the dir we write stuff to, then we should not write more data there i guess..

Reply:
ok, will add a check in the next push